File: competitive-programming/leet-code/413/numberOfArithmeticSlices.js
var numberOfArithmeticSlices = function (nums) {
    if (nums.length < 3) return 0;
    const diffs = [];
    for (let i = 0; i < nums.length - 1; i++) {
        diffs.push(nums[i + 1] - nums[i]);
    }

    let count = 0;
    let start = 0;
    while (start < diffs.length) {
        let end = start;
        while (end + 1 < diffs.length && diffs[end + 1] === diffs[start]) {
            end++;
        }
        const len = end - start + 1;
        if (len >= 2) count += (len * (len - 1)) / 2;
        start = end + 1;
    }

    return count;
};

File: competitive-programming/leet-code/413/number_of_arithmetic_slices.py
from typing import List

def number_of_arithmetic_slices(nums: List[int]) -> int:
    if len(nums) < 3:
        return 0

    diffs = []
    for i in range(len(nums) - 1):
        diffs.append(nums[i + 1] - nums[i])

    count = 0
    start = 0

    while start < len(diffs):
        end = start
        while end + 1 < len(diffs) and diffs[end + 1] == diffs[start]:
            end += 1

        length = end - start + 1

        if length >= 2:
            count += (length * (length - 1)) // 2

        start = end + 1

    return count